src/wallet/wallet.cpp

```diff
@@ -2938,26 +2938,29 @@ bool CWallet::CreateCoinStake(
     CScript scriptPubKeyKernel;
     bool fKernelFound = false;
     int nAttempts = 0;
-    for (const auto& out : *availableCoins) {
-        COutPoint outPoint = COutPoint(out.tx->GetHash(), out.i);
-        CPivStake stakeInput(out.tx->vout[out.i],
+    for (auto it = availableCoins->begin(); it != availableCoins->end();) {
+        COutPoint outPoint = COutPoint(it->tx->GetHash(), it->i);
+        CPivStake stakeInput(it->tx->vout[it->i],
                              outPoint,
-                             out.pindex);
+                             it->pindex);
 
         //new block came in, move on
         if (WITH_LOCK(cs_main, return chainActive.Height()) != pindexPrev->nHeight) return false;
 
         // Make sure the wallet is unlocked and shutdown hasn't been requested
         if (IsLocked() || ShutdownRequested()) return false;
 
-        // Make sure the stake input hasn't been spent in the meantime
+        // Make sure the stake input hasn't been spent since last check
         if (IsSpent(outPoint)) {
+            // remove it from the available coins
+            it = availableCoins->erase(it);
             continue;
         }
 
         // This should never happen
         if (stakeInput.IsZPIV()) {
             LogPrintf("%s: ERROR - zPOS is disabled\n", __func__);
+            it++;
```

Comment:
we can remove this check, the stake input is created at the beginning of the loop, `IsZPIV()` is always false.

Reply:
Out of scope here, but sure.

Reply:
yep, wrote it on the main comment. No need to include it here.